pallets/rc-migrator/src/scheduler.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+use frame_support::traits::schedule::v3::TaskName;
+use pallet_scheduler::{RetryConfig, TaskAddress};
+
+/// Stage of the scheduler pallet migration.
+#[derive(Encode, Decode, Clone, Default, RuntimeDebug, TypeInfo, MaxEncodedLen, PartialEq, Eq)]
+pub enum SchedulerStage<BlockNumber> {
+	#[default]
+	IncompleteSince,
+	Agenda(Option<BlockNumber>),
+	Retries(Option<TaskAddress<BlockNumber>>),
+	Lookup(Option<TaskName>),
+	Finished,
+}
+
+/// Message that is being sent to the AH Migrator.
+#[derive(Encode, Decode, Debug, Clone, TypeInfo, MaxEncodedLen, PartialEq, Eq)]
+pub enum RcSchedulerMessage<BlockNumber, Scheduled> {
+	IncompleteSince(BlockNumber),
+	Agenda((BlockNumber, Vec<Option<Scheduled>>)),
+	Retries((TaskAddress<BlockNumber>, RetryConfig<BlockNumber>)),
+	Lookup((TaskName, TaskAddress<BlockNumber>)),
+}
+
+pub type RcSchedulerMessageOf<T> = RcSchedulerMessage<BlockNumberFor<T>, alias::ScheduledOf<T>>;
+
+pub struct SchedulerMigrator<T: Config> {
+	_phantom: PhantomData<T>,
+}
+
+impl<T: Config> PalletMigration for SchedulerMigrator<T> {
+	type Key = SchedulerStage<BlockNumberFor<T>>;
+	type Error = Error<T>;
+	fn migrate_many(
+		last_key: Option<Self::Key>,
+		weight_counter: &mut WeightMeter,
+	) -> Result<Option<Self::Key>, Self::Error> {
+		let mut last_key = last_key.unwrap_or(SchedulerStage::IncompleteSince);
+		let mut messages = Vec::new();
+
```

Comment:
We also have to go through all pallets `on_initialize` code and possibly disable it during the migration. Otherwise some pallets may break from partial state being migrated. I will add it as TODO

Reply:
Right. I have an idea how to do it for the scheduler